### DISCUSSION TITLE
Installing Docker on Pengwin without Docker Desktop

### DISCUSSION THREAD

**Discussion by avinashupadhya99** [WhitewaterFoundry/Pengwin#734] [Q&A] [answered]
Hello, I am trying to install docker on Pengwin without Docker Desktop. Is the process documented anywhere?
 
 I would really appreciate any help with this. Thanks!

**Comment by crramirez** [comment]
Hello @avinashupadhya99,

No. I'll assemble one and paste it here. Maybe then we can add this to pengwin-setup.

Regards,
Carlos

**Comment by crramirez** [comment, marked as answer]
### How to install docker in Pengwin without Docker Desktop

```bash
sudo apt update
sudo apt install -y  --no-install-recommends gnupg2
```
```bash
source /etc/os-release
curl -fsSL https://download.docker.com/linux/${ID}/gpg | sudo apt-key add -
echo "deb [arch=amd64] https://download.docker.com/linux/${ID} ${VERSION_CODENAME} stable" | sudo tee /etc/apt/sources.list.d/docker.list
sudo apt update
```
```bash
sudo apt install -y docker-ce docker-ce-cli containerd.io
```
```bash
sudo usermod -aG docker $USER
```
```bash
DOCKER_DIR=/mnt/wsl/shared-docker
mkdir -pm o=,ug=rwx "$DOCKER_DIR"
sudo chgrp docker "$DOCKER_DIR"
sudo mkdir /etc/docker
```
```bash
sudo <your_text_editor> /etc/docker/daemon.json
```
```json
{
   "hosts": ["unix:///mnt/wsl/shared-docker/docker.sock"],
   "iptables": false
}
```
```bash
sudo <your_text_editor> /etc/profile.d/docker.sh
```
```sh
#! bin/sh

DOCKER_DISTRO="${WSL_DISTRO_NAME}"
DOCKER_DIR=/mnt/wsl/shared-docker
DOCKER_SOCK="$DOCKER_DIR/docker.sock"
export DOCKER_HOST="unix://$DOCKER_SOCK"

connected=$(docker version 2>&1 | grep -c "daemon\|error")
if [ ${connected} != 0  ]; then
  rm -f "${DOCKER_SOCK}"
fi

unset connected

if [ ! -S "$DOCKER_SOCK" ]; then
   mkdir -pm o=,ug=rwx "$DOCKER_DIR"
   chgrp docker "$DOCKER_DIR"
   /mnt/c/Windows/System32/wsl.exe -d $DOCKER_DISTRO sh -c "nohup sudo -b dockerd </dev/null > $DOCKER_DIR/dockerd.log 2>&1"
fi
```
```bash
echo "%docker ALL=(ALL) NOPASSWD: /usr/bin/dockerd" | sudo EDITOR='tee -a' visudo --quiet --file=/etc/sudoers.d/dockerd
```
```bash
COMPOSE_VERSION=1.29.2
sudo curl -L "https://github.com/docker/compose/releases/download/$COMPOSE_VERSION/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
sudo chmod +x /usr/local/bin/docker-compose
```

Restart Pengwin

### Steps from
https://dev.solita.fi/2021/12/21/docker-on-wsl2-without-docker-desktop.html

**Reply by avinashupadhya99** [reply]
Thank you! I have 2 errors with the startup script. I was able to verify that I can manually start `dockerd` and run a docker command in another window. But the startup script errors out with - 
```
chgrp: changing group of '/mnt/wsl/shared-docker': Operation not permitted
bash: /mnt/c/Windows/System32/wsl.exe: No such file or directory
```
When I run `ls /mnt/c/`, it returns nothing. Looks like my C drive is not mounted. Can you please let me know how I can mount it and also fix the permission issue with chgrp? Thanks

**Reply by crramirez** [reply]
Ok probably you have installed Docker in pengwin-setup. Deinstall it first: `pengwin-setup uninstall DOCKER`

Maybe C: is mounted in /c/ instead of /mnt/c. You can do one of two things:

1. Change everything in the above script from **/mnt/** to **/**: /mnt/c -> /c, /mnt/wsl -> /wsl
2. Edit /etc/wsl.conf and change the root back to /mnt: `sudo sed -i 's$\(root=\)\(.*\)$\1/mnt$' /etc/wsl.conf` and run `wsl --terminate $WSL_DISTRO_NAME`

**Reply by avinashupadhya99** [reply]
Thank you, it worked!

**Reply by C0coleplub0** [reply]
It worked as a charm thx a lot dude @crramirez

**Reply by sharpninja** [reply]
Can this be added to pengwin-setup

**Reply by crramirez** [reply]
It is a lot of work making it right. But as you asked for it you raised it's priority. So, yes it will be added to pengwin-setup, I hope this year

**Reply by mikkop71** [reply]
Any news on the schedule?

**Reply by crramirez** [reply]
We are working on that, but still we don't have a release date
